File: src/main/java/me/testaccount666/serversystem/commands/executables/unlimited/ListenerUnlimited.java
package me.testaccount666.serversystem.commands.executables.unlimited;

import me.testaccount666.serversystem.ServerSystem;
import me.testaccount666.serversystem.annotations.RequiredCommands;
import me.testaccount666.serversystem.commands.interfaces.ServerSystemCommandExecutor;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.block.Container;
import org.bukkit.block.data.Directional;
import org.bukkit.entity.*;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.block.BlockDispenseArmorEvent;
import org.bukkit.event.block.BlockDispenseEvent;
import org.bukkit.event.block.BlockPlaceEvent;
import org.bukkit.event.entity.CreatureSpawnEvent;
import org.bukkit.event.entity.ItemSpawnEvent;
import org.bukkit.event.player.*;
import org.bukkit.inventory.InventoryHolder;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.AxolotlBucketMeta;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.inventory.meta.TropicalFishBucketMeta;

import javax.annotation.Nullable;
import java.util.Set;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Listener that handles events related to unlimited items functionality.
 * An unlimited item is an item that doesn't get consumed or reduced in quantity when used.
 */
@RequiredCommands(requiredCommands = CommandUnlimited.class)
public class ListenerUnlimited implements Listener {
    private CommandUnlimited _commandUnlimited;

    /**
     * Checks if this listener can be registered based on the available commands.
     *
     * @param requiredCommands Set of available command executors
     * @return true if the required CommandUnlimited is available, false otherwise
     */
    public boolean canRegister(Set<ServerSystemCommandExecutor> requiredCommands) {
        var canRegister = new AtomicBoolean(false);

        requiredCommands.forEach(command -> {
            if (!(command instanceof CommandUnlimited commandUnlimited)) return;

            _commandUnlimited = commandUnlimited;
            canRegister.set(true);
        });

        return canRegister.get();
    }

    /**
     * Handles consumption of unlimited items, preventing them from being consumed.
     */
    @EventHandler
    public void onConsume(PlayerItemConsumeEvent event) {
        var itemInHand = event.getItem();
        if (!isUnlimited(itemInHand)) return;

        event.setReplacement(itemInHand);
    }

    /**
     * Handles placement of unlimited blocks, ensuring they don't get consumed.
     */
    @EventHandler
    public void onBlockPlace(BlockPlaceEvent event) {
        var itemInHand = event.getItemInHand();
        if (!isUnlimited(itemInHand)) return;

        itemInHand.setAmount(itemInHand.getAmount() + 1);
    }

    /**
     * Handles dropping of unlimited items, creating a non-unlimited copy.
     */
    @EventHandler
    public void onItemDrop(PlayerDropItemEvent event) {
        var itemInHand = event.getItemDrop().getItemStack();
        if (!isUnlimited(itemInHand)) return;

        event.setCancelled(true);
        var newItem = event.getItemDrop().getWorld().dropItemNaturally(event.getItemDrop().getLocation(), itemInHand);

        var newItemStack = newItem.getItemStack();
        var newItemMeta = newItemStack.getItemMeta();
        var newDataContainer = newItemMeta.getPersistentDataContainer();
        newDataContainer.remove(_commandUnlimited.unlimitedKey);
        newItemStack.setItemMeta(newItemMeta);
    }

    /**
     * Handles bucket emptying for unlimited buckets.
     */
    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    public void onBucketEmpty(PlayerBucketEmptyEvent event) {
        var itemInHand = event.getPlayer().getInventory().getItem(event.getHand());
        if (!isUnlimited(itemInHand)) return;

        event.setCancelled(true);
        handleBucketEmpty(event);
    }

    /**
     * Helper method to handle different types of bucket emptying.
     */
    private void handleBucketEmpty(PlayerBucketEmptyEvent event) {
        var block = event.getBlock();
        var location = block.getLocation().add(0.5, 0.5, 0.5);

        switch (event.getBucket()) {
            case WATER_BUCKET -> location.getWorld().setType(location, Material.WATER);
            case LAVA_BUCKET -> location.getWorld().setType(location, Material.LAVA);
            case POWDER_SNOW_BUCKET -> location.getWorld().setType(location, Material.POWDER_SNOW);
            case AXOLOTL_BUCKET, COD_BUCKET, SALMON_BUCKET, TADPOLE_BUCKET,
                 PUFFERFISH_BUCKET, TROPICAL_FISH_BUCKET -> handleAquaticBucketEmpty(event, location);
        }
    }

    /**
     * Helper method to handle emptying of aquatic mob buckets.
     */
    private void handleAquaticBucketEmpty(PlayerBucketEmptyEvent event, Location location) {
        location.getWorld().setType(location, Material.WATER);
        var itemInHand = event.getPlayer().getInventory().getItem(event.getHand());
        var itemMeta = itemInHand.getItemMeta();

        switch (event.getBucket()) {
            case AXOLOTL_BUCKET -> location.getWorld().spawn(location, Axolotl.class,
                    CreatureSpawnEvent.SpawnReason.BUCKET, axolotl -> handleEntitySpawn(axolotl, itemMeta));
            case COD_BUCKET -> location.getWorld().spawn(location, Cod.class,
                    CreatureSpawnEvent.SpawnReason.BUCKET, cod -> handleEntitySpawn(cod, itemMeta));
            case SALMON_BUCKET -> location.getWorld().spawn(location, Salmon.class,
                    CreatureSpawnEvent.SpawnReason.BUCKET, salmon -> handleEntitySpawn(salmon, itemMeta));
            case TADPOLE_BUCKET -> location.getWorld().spawn(location, Tadpole.class,
                    CreatureSpawnEvent.SpawnReason.BUCKET, tadpole -> handleEntitySpawn(tadpole, itemMeta));
            case PUFFERFISH_BUCKET -> location.getWorld().spawn(location, PufferFish.class,
                    CreatureSpawnEvent.SpawnReason.BUCKET, pufferfish -> handleEntitySpawn(pufferfish, itemMeta));
            case TROPICAL_FISH_BUCKET -> location.getWorld().spawn(location, TropicalFish.class,
                    CreatureSpawnEvent.SpawnReason.BUCKET, tropicalfish -> handleEntitySpawn(tropicalfish, itemMeta));
        }
    }

    /**
     * Handles entity spawning from buckets, applying metadata from the bucket.
     */
    private void handleEntitySpawn(Entity entity, ItemMeta itemMeta) {
        if (itemMeta instanceof TropicalFishBucketMeta bucketMeta && entity instanceof TropicalFish tropicalFish) if (bucketMeta.hasVariant()) {
            tropicalFish.setPattern(bucketMeta.getPattern());
            tropicalFish.setBodyColor(bucketMeta.getBodyColor());
            tropicalFish.setPatternColor(bucketMeta.getPatternColor());
        }

        if (itemMeta instanceof AxolotlBucketMeta bucketMeta && entity instanceof Axolotl axolotl)
            if (bucketMeta.hasVariant()) axolotl.setVariant(bucketMeta.getVariant());

        if (itemMeta.hasCustomName()) entity.customName(itemMeta.customName());
    }

    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    public void onBucketFill(PlayerBucketFillEvent event) {
        var itemInHand = event.getPlayer().getInventory().getItem(event.getHand());
        if (!isUnlimited(itemInHand)) return;

        event.setCancelled(true);

        var block = event.getBlock();
        var location = block.getLocation();

        location.getWorld().setType(location, Material.AIR);
    }

    @EventHandler
    public void onItemDamage(PlayerItemDamageEvent event) {
        var itemInHand = event.getItem();
        if (!isUnlimited(itemInHand)) return;

        event.setCancelled(true);
    }

    @EventHandler
    public void onDispenserFire(BlockDispenseEvent event) {
        var itemInHand = event.getItem();

        if (!isUnlimited(itemInHand)) return;

        var dispenserBlock = event.getBlock();
        var blockState = dispenserBlock.getState();
        if (!(blockState instanceof Container container)) return;

        var blockData = dispenserBlock.getBlockData();
        if (!(blockData instanceof Directional directional)) return;
        var velocity = event.getVelocity();

        var facing = directional.getFacing();
        var facingBlock = dispenserBlock.getRelative(facing);

        var facingX = facingBlock.getX();
        var facingY = facingBlock.getY();
        var facingZ = facingBlock.getZ();

        var velocityX = velocity.getX();
        var velocityY = velocity.getY();
        var velocityZ = velocity.getZ();

        // Very quirky: when placing a block, the velocity is faked as the block’s coordinates.
        // When dropping an item, velocity is an actual direction vector.
        // Additionally, `BlockDispenseEvent` is fired AFTER removing the item from inventory, so we cannot grab the index
        if (facingX != velocityX || facingY != velocityY || facingZ != velocityZ) {
            Bukkit.getScheduler().runTaskLater(ServerSystem.Instance, () -> container.getInventory().addItem(itemInHand), 1L);
            return;
        }

        var foundIndex = findSimilarItemIndex(itemInHand, container);
        if (foundIndex == -1) return;

        Bukkit.getScheduler().runTaskLater(ServerSystem.Instance, () -> container.getInventory().setItem(foundIndex, itemInHand), 1L);
    }

    private int findSimilarItemIndex(ItemStack comparingItem, InventoryHolder inventoryHolder) {
        var foundIndex = -1;

        var maxIndex = inventoryHolder.getInventory().getSize();

        for (var index = 0; index < maxIndex; index++) {
            var itemStack = inventoryHolder.getInventory().getItem(index);
            if (itemStack == null) continue;
            if (!itemStack.isSimilar(comparingItem)) continue;

            foundIndex = index;
        }

        return foundIndex;
    }

    @EventHandler
    public void onDispenserArmorFire(BlockDispenseArmorEvent event) {
        var itemInHand = event.getItem();
        if (!isUnlimited(itemInHand)) return;

        var newItemStack = itemInHand.clone();
        var newItemMeta = newItemStack.getItemMeta();
        var newDataContainer = newItemMeta.getPersistentDataContainer();

        newDataContainer.remove(_commandUnlimited.unlimitedKey);
        newItemStack.setItemMeta(newItemMeta);

        event.setItem(newItemStack);
    }

    /**
     * Checks if an item is marked as unlimited.
     *
     * @param itemStack The item to check
     * @return true if the item is unlimited, false otherwise
     */
    private boolean isUnlimited(@Nullable ItemStack itemStack) {
        if (itemStack == null) return false;

        var itemMeta = itemStack.getItemMeta();
        if (itemMeta == null) return false;

        var dataContainer = itemMeta.getPersistentDataContainer();
        return dataContainer.has(_commandUnlimited.unlimitedKey);
    }

    @EventHandler
    public void onItemSpawn(ItemSpawnEvent event) {
        var itemStack = event.getEntity().getItemStack();
        if (!isUnlimited(itemStack)) return;

        var itemMeta = itemStack.getItemMeta();
        var dataContainer = itemMeta.getPersistentDataContainer();

        dataContainer.remove(_commandUnlimited.unlimitedKey);
        itemStack.setItemMeta(itemMeta);

        event.getEntity().setItemStack(itemStack);
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/commands/executables/unlimited/ListenerUnlimited.kt
package me.testaccount666.serversystem.commands.executables.unlimited

import io.papermc.paper.datacomponent.DataComponentTypes
import io.papermc.paper.datacomponent.item.UseRemainder.useRemainder
import me.testaccount666.paperktx.extensions.*
import me.testaccount666.paperktx.scheduler.skedule.okkero.schedule
import me.testaccount666.serversystem.ServerSystem.Companion.instance
import me.testaccount666.serversystem.annotations.RequiredCommands
import me.testaccount666.serversystem.commands.interfaces.ServerSystemCommandExecutor
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.block.Container
import org.bukkit.block.data.Directional
import org.bukkit.entity.*
import org.bukkit.event.*
import org.bukkit.event.block.*
import org.bukkit.event.entity.ItemSpawnEvent
import org.bukkit.event.player.*
import org.bukkit.inventory.InventoryHolder
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.*

/**
 * Listener that handles events related to unlimited items functionality.
 * An unlimited item is an item that doesn't get consumed or reduced in quantity when used.
 */
@RequiredCommands([CommandUnlimited::class])
class ListenerUnlimited : Listener {
    private lateinit var _commandUnlimited: CommandUnlimited

    /**
     * Checks if this listener can be registered based on the available commands.
     *
     * @param requiredCommands Set of available command executors
     * @return true if the required CommandUnlimited is available, false otherwise
     */
    fun canRegister(requiredCommands: Set<ServerSystemCommandExecutor>): Boolean {
        _commandUnlimited = requiredCommands.firstNotNullOfOrNull { it as? CommandUnlimited } ?: return false
        return true
    }

    /**
     * Handles consumption of unlimited items, preventing them from being consumed.
     */
    @EventHandler
    fun onConsume(event: PlayerItemConsumeEvent) {
        val itemInHand = event.item.takeIf(::isUnlimited) ?: return
        event.replacement = itemInHand
    }

    /**
     * Handles placement of unlimited blocks, ensuring they don't get consumed.
     */
    @EventHandler
    fun onBlockPlace(event: BlockPlaceEvent) {
        val itemInHand = event.itemInHand.takeIf(::isUnlimited) ?: return
        itemInHand.setData(DataComponentTypes.USE_REMAINDER, useRemainder(itemInHand))
    }

    /**
     * Handles dropping of unlimited items, creating a non-unlimited copy.
     */
    @EventHandler
    fun onItemDrop(event: PlayerDropItemEvent) {
        val itemInHand = event.itemDrop.itemStack.takeIf(::isUnlimited) ?: return

        event.isCancelled = true
        event.itemDrop.world.dropItem(event.itemDrop.location, itemInHand) { item ->
            item.velocity = event.itemDrop.velocity
            item.itemStack.editPersistentDataContainer { it.remove(_commandUnlimited.unlimitedKey) }
        }
    }

    /**
     * Handles bucket emptying for unlimited buckets.
     */
    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    fun onBucketEmpty(event: PlayerBucketEmptyEvent) {
        event.getPlayer().inventory[event.hand].takeIf(::isUnlimited) ?: return

        event.isCancelled = true
        handleBucketEmpty(event)
    }

    /**
     * Helper method to handle different types of bucket emptying.
     */
    private fun handleBucketEmpty(event: PlayerBucketEmptyEvent) {
        val block = event.block
        val location = block.location.add(0.5, 0.5, 0.5)

        when (event.bucket) {
            Material.WATER_BUCKET -> location.world.setType(location, Material.WATER)
            Material.LAVA_BUCKET -> location.world.setType(location, Material.LAVA)
            Material.POWDER_SNOW_BUCKET -> location.world.setType(location, Material.POWDER_SNOW)
            Material.AXOLOTL_BUCKET, Material.COD_BUCKET, Material.SALMON_BUCKET,
            Material.TADPOLE_BUCKET, Material.PUFFERFISH_BUCKET, Material.TROPICAL_FISH_BUCKET,
                -> handleAquaticBucketEmpty(event, location)

            else -> Unit
        }
    }

    /**
     * Helper method to handle emptying of aquatic mob buckets.
     */
    private fun handleAquaticBucketEmpty(event: PlayerBucketEmptyEvent, location: Location) {
        location.world.setType(location, Material.WATER)
        val itemInHand = event.getPlayer().inventory[event.hand]
        val itemMeta = itemInHand.itemMeta

        when (event.bucket) {
            Material.AXOLOTL_BUCKET -> location.spawn<Axolotl> { handleEntitySpawn(this, itemMeta) }
            Material.COD_BUCKET -> location.spawn<Cod> { handleEntitySpawn(this, itemMeta) }
            Material.SALMON_BUCKET -> location.spawn<Salmon> { handleEntitySpawn(this, itemMeta) }
            Material.TADPOLE_BUCKET -> location.spawn<Tadpole> { handleEntitySpawn(this, itemMeta) }
            Material.PUFFERFISH_BUCKET -> location.spawn<PufferFish> { handleEntitySpawn(this, itemMeta) }
            Material.TROPICAL_FISH_BUCKET -> location.spawn<TropicalFish> { handleEntitySpawn(this, itemMeta) }

            else -> Unit
        }
    }

    /**
     * Handles entity spawning from buckets, applying metadata from the bucket.
     */
    private fun handleEntitySpawn(entity: Entity, itemMeta: ItemMeta) {
        if ((itemMeta is TropicalFishBucketMeta && entity is TropicalFish) && itemMeta.hasVariant()) {
            entity.pattern = itemMeta.pattern
            entity.bodyColor = itemMeta.bodyColor
            entity.patternColor = itemMeta.patternColor
        }

        if ((itemMeta is AxolotlBucketMeta && entity is Axolotl) && itemMeta.hasVariant()) entity.variant = itemMeta.variant

        if (itemMeta.hasCustomName()) entity.customName(itemMeta.customName())
    }

    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    fun onBucketFill(event: PlayerBucketFillEvent) {
        event.getPlayer().inventory[event.hand].takeIf(::isUnlimited) ?: return

        event.isCancelled = true
        event.block.location.run { world.setType(this, Material.AIR) }
    }

    @EventHandler
    fun onItemDamage(event: PlayerItemDamageEvent) {
        event.item.takeIf(::isUnlimited) ?: return
        event.isCancelled = true
    }

    @EventHandler
    fun onDispenserFire(event: BlockDispenseEvent) {
        val itemInHand = event.item.takeIf(::isUnlimited) ?: return

        val dispenserBlock = event.getBlock()
        val blockState = dispenserBlock.state as? Container ?: return

        val blockData = dispenserBlock.blockData as? Directional ?: return

        val facing = blockData.facing
        val facingBlock = dispenserBlock.getRelative(facing)

        val facingX = facingBlock.x
        val facingY = facingBlock.y
        val facingZ = facingBlock.z

        val (velocityX, velocityY, velocityZ) = event.velocity

        instance.schedule {
            // Very quirky: when placing a block, the velocity is faked as the block’s coordinates.
            // When dropping an item, velocity is an actual direction vector.
            // Additionally, `BlockDispenseEvent` is fired AFTER removing the item from inventory, so we cannot grab the index
            if (facingX.toDouble() != velocityX || facingY.toDouble() != velocityY || facingZ.toDouble() != velocityZ) {
                waitFor(1L)
                blockState.inventory.addItem(itemInHand)
                return@schedule
            }

            val foundIndex = findSimilarItemIndex(itemInHand, blockState)
            if (foundIndex == -1) return@schedule

            waitFor(1L)
            blockState.inventory[foundIndex] = itemInHand
        }
    }

    private fun findSimilarItemIndex(comparingItem: ItemStack?, inventoryHolder: InventoryHolder): Int {
        var foundIndex = -1

        val maxIndex = inventoryHolder.inventory.size

        for (index in 0..<maxIndex) {
            val itemStack = inventoryHolder.inventory[index].takeUnless { it.isAir() } ?: continue
            if (!itemStack.isSimilar(comparingItem)) continue

            foundIndex = index
        }

        return foundIndex
    }

    @EventHandler
    fun onDispenserArmorFire(event: BlockDispenseArmorEvent) {
        val itemInHand = event.item.takeIf(::isUnlimited) ?: return

        event.item = itemInHand.clone().edit { persistentDataContainer.remove(_commandUnlimited.unlimitedKey) }
    }

    /**
     * Checks if an item is marked as unlimited.
     *
     * @param itemStack The item to check
     * @return true if the item is unlimited, false otherwise
     */
    private fun isUnlimited(itemStack: ItemStack?): Boolean {
        val itemMeta = itemStack?.itemMeta ?: return false

        return itemMeta.persistentDataContainer.has(_commandUnlimited.unlimitedKey)
    }

    @EventHandler
    fun onItemSpawn(event: ItemSpawnEvent) {
        val itemStack = event.getEntity().itemStack.takeIf(::isUnlimited) ?: return

        event.getEntity().itemStack = itemStack.edit { persistentDataContainer.remove(_commandUnlimited.unlimitedKey) }
    }
}